### DISCUSSION TITLE
Create a custom page showing chosen content?

### DISCUSSION THREAD

**Discussion by miggland** [inventree/InvenTree#4260] [General]
I want to keep things as simple as possible for the guys in the warehouse doing receiving, packing etc.

Originally, I was thinking of creating a report for them to look at, with pictures and maybe QR codes for receiving items. However, it seems unnecessary to go the paper-based route. I would like to create a page, which looks like a report, containing for example only a certain PO and the items expected. Possibly with buttons to receive a line/all lines.

What is the best way to do this? Can I use the plugin functionality for this, or would I have to host these pages separately and get data through the API?

(I'm not sure the URL plugin mixin helps here, I don't quite understand what that is useful for..)

**Comment by wolflu05** [comment]
Yes, you can do this with plugins. What mixin you should use depends on where you want to add your custom page. If you just want to have another panel in the list, you can use the panels mixin, but if you want to have a separate page you can use the urls mixin which gives you the ability to serve own templates through a url. You can then e.g. use the navigations mixin to add a link to inventrees navigation bar to reference to your custom page, or just use the panels mixin for that reference.

**Reply by miggland** [reply]
Ah, then I'm probably going to go with the URLs plugin. Having a completely different page seems like what I want (even though I might change my mind in the future depending on how it's received.. ;) )

Thanks, I'll give it a shot..

**Reply by miggland** [reply]
Alright, I'm slowly getting there! This will work out in the end :)

**Reply by matmair** [reply]
@miggland I might be a little late but I actually made a cookie cutter to make starting plugins easier: https://github.com/matmair/inventree-meta-plugin#cookie-cutter

**Comment by miggland** [comment]
I made a plugin doing what I want, using the test system, which worked well. This is on the newest master branch (dev. server).

Now, I have the problem that I can't get it to work on my main system, running 0.9.2. It's installed, but when I try to open the URL (which is listed on the plugin page), I get redirected to /index/.

Any ideas? I can't even see error messages..

This is the main part of the plugin:

```
"""
    CustomPager plugin
    
    Creates custom pages, with a separate template, for carrying out simple tasks/showing interactive reports.
    
    This is used to show users a simpler, custom interface.
"""

from django.utils.translation import gettext_lazy as _
from django.urls import re_path
from django.http import HttpResponse
from django.template import loader

import order.models as models
from InvenTree.status_codes import PurchaseOrderStatus

from plugin import InvenTreePlugin
from plugin.mixins import UrlsMixin


def start_page(request):

    template = loader.get_template('custompager_start.html')

    # Get all pending orders
    purchase_orders = models.PurchaseOrder.objects.filter(status=PurchaseOrderStatus.PLACED)

    context = {
        'purchase_orders': purchase_orders,
    }

    return HttpResponse(template.render(context, request))


def po_detail(request, pk):

    template = loader.get_template('custompager_purchase_order_detail.html')

    order = models.PurchaseOrder.objects.get(pk=pk)

    context = {
        'pk': pk,
        'order': order,
    }

    return HttpResponse(template.render(context, request))


class CustomPagerPlugin(UrlsMixin, InvenTreePlugin):
    """Plugin to display custom pages"""

    NAME = "CustomPagerPlugin"
    SLUG = "cpager"
    TITLE = _("Custom Pager Plugin")
    DESCRIPTION = _("Shows custom pages at selected URLs")
    VERSION = "0.1.1"
    AUTHOR = _("")

    URLS = [
        # PO Detail view
        re_path(r'po/(?P<pk>\d+)/', po_detail, name='custompager-po-detail'),
        
        # Start page
        re_path(r'^.*$', start_page, name='custompager-start'),
    ]
```

**Comment by miggland** [comment]
In fact, nothing works. I have enabled all the plugin integrations on the settings page, and enabled the plugins.

-> An even simpler version than above (https://github.com/inventree/inventree-docs/pull/432) doesn't work
-> The "sample.py" plugin from the source doesn't work

What's could be wrong? This is annoying, as I was hoping to just install this plugin on my production server and start using it. Now I'm stuck :(

**Reply by wolflu05** [reply]
Have you tried restarting inventree?

**Reply by miggland** [reply]
I do have this in my logs since a few (failed) attempts at using `git+` type URLs/packages of my own plugins:
`
fatal: not a git repository (or any of the parent directories): .git
`

These won't go away, even after removing the failed plugins, re-installing and applying a backup. Is there a way to clean out the plugin installation to get rid of the old trash which seems to be littering?

**Reply by miggland** [reply]
> Have you tried restarting inventree?

Only about 100 times.

**Reply by wolflu05** [reply]
What is the link on what you click look like?

**Reply by miggland** [reply]
![image](https://user-images.githubusercontent.com/1314853/215776075-81fa1145-a50a-4986-abcf-356217a978ed.png)

Either the link top left, or one of the two buttons on the right. The links are what I expect..
They lead to (changed domain name):
```
https://inventory.mycompany.eu/plugin/cpager/
https://inventory.mycompany.eu/plugin/cpager/.*$
https://inventory.mycompany.eu/plugin/cpager/po/(?P%3Cpk%3E\d+)/
```

Respectively. The last two don't really make sense. However, if I enter the links as they should be manually, I get the same effect.
```
https://inventory.mycompany.eu/plugin/cpager/
https://inventory.mycompany.eu/plugin/cpager/po/1/
```

All of these links are redirected to this page:
```
https://inventory.mycompany.eu/index/
```

With the sample plugin, the URL was different of course. However, the results are the same - the URLs are not found.

This may be the same as described here:
https://github.com/inventree/InvenTree/issues/4052

For which there is a patch:
https://github.com/inventree/InvenTree/pull/4065

This patch unfortunately didn't make it into 0.9.2..

**Reply by miggland** [reply]
Alright, was able to fix it.. I added the following code:
```
    def setup_urls(self):
        """Urls that are exposed by this plugin."""
        return [
            # PO Detail view
            re_path(r'po/(?P<pk>\d+)/', po_detail, name='custompager-po-detail'),

            # Start page
            re_path(r'^.*$', start_page, name='custompager-start'),
        ]
```
to my plugin. In the newer version of Inventree, this is not required (defining variable URLS) seems to be enough. Another restart, and now the the URLs work.

**Reply by matmair** [reply]
That is already patched in master. Rewriting the setup_urls replaced the broken codes function in 0.9.x.

**Comment by matmair** [comment]
This is also being implemented as a navigation plugin https://github.com/inventree/InvenTree/pull/12900
